test/fft_test.hs

```diff
@@ -20,35 +21,48 @@ import Control.Applicative
 import Control.Arrow
 import Control.Monad (forM_, unless)
 import Data.Complex
-import Data.Foldable (sum)
+import Data.Foldable (Foldable, sum, foldl')
 import Data.Newtypes.PrettyDouble
 import System.Exit (exitFailure)
-import TypeUnary.Nat (IsNat, natToZ, Nat(..), nat, N2, N3, N4, N5)  -- , N6)
+import TypeUnary.Nat (IsNat, Nat(..), nat, N2, N3, N4, N5)  -- , N6)
 
 -- import Test.QuickCheck (choose, vectorOf, elements, collect)
 import Test.QuickCheck (choose, vectorOf)
 import Test.QuickCheck.Arbitrary
 import Test.QuickCheck.All (quickCheckAll)
 
-import Circat.Scan (scanlTEx)
+import Circat.Scan (lproducts, LScan)
 import qualified Circat.Pair as P
 import qualified Circat.RTree as RT
 import Circat.RTree (bottomSplit)
 
 type RTree = RT.Tree
 
--- Phasor, as a function of tree depth.
-phasor :: (IsNat n, RealFloat a, Enum a) => Nat n -> RTree n (Complex a)
-phasor n = scanlTEx (*) 1 (pure phaseDelta)
-    where phaseDelta = cis ((-pi) / 2 ** natToZ n)
-
--- Radix-2, DIT FFT
-fft_r2_dit :: (IsNat n, RealFloat a, Enum a) => RTree n (Complex a) -> RTree n (Complex a)
-fft_r2_dit = fft_r2_dit' nat
-
-fft_r2_dit' :: (RealFloat a, Enum a) => Nat n -> RTree n (Complex a) -> RTree n (Complex a)
-fft_r2_dit'  Zero    = id
-fft_r2_dit' (Succ n) = RT.toB . P.inP (uncurry (+) &&& uncurry (-)) . P.secondP (liftA2 (*) (phasor n)) . fmap (fft_r2_dit' n) . bottomSplit
+-- FFT, as a class
+--   The LScan constraint comes from the use of 'lproducts', in 'phasor'.
+class (LScan f) => FFT f where
+    fft  :: (RealFloat a) => f (Complex a) -> f (Complex a)
+
+instance IsNat n => FFT (RTree n) where
+    fft = fft' nat
+        where   fft' :: (RealFloat a) => Nat n -> RTree n (Complex a) -> RTree n (Complex a)
+                fft' Zero     = id
+                fft' (Succ n) = inDIT $ fftP . P.secondP addPhase . fmap (fft' n)
+                    where   inDIT g  = RT.toB . g . bottomSplit
+                            fftP     = P.inP (uncurry (+) &&& uncurry (-))
+                            addPhase = liftA2 (*) id phasor
+
+-- Phasor, as a general function on LScans.
+--   Gives the "length" (i.e. - number of elements in) of a Foldable.
+--   (Soon, to be provided by the Foldable class, as "length".)
+flen :: (Foldable f) => f a -> Int
+flen = foldl' (flip ((+) . const 1)) 0
+
+--   Given a Foldable Applicative LScan, construct its matching phasor.
+phasor :: (Applicative f, Foldable f, LScan f, RealFloat b) => f a -> f (Complex b)
```

Comment:
```haskell
flen :: (Foldable f) => f a -> Int
flen = foldl' (flip ((+) . const 1)) 0
```

More simply and more parallel-friendly:

```haskell
flen = sum . fmap (const 1)
```

For types like `RTree n`, however, you don't need an actual tree, since the type determines the length. In `phasor`, you could write

```haskell
n = sum (pure 1 :: f Int)
```

Then you can drop the `f a` argument. You'll have to add the `ScopedTypeVariables` pragma and add and explicit universal quantifier for `f` and `b`. (You'll then also have to rename "`n`" in the signature for `fft'`.)

```haskell
phasor :: forall f a. (Applicative f, Foldable f, LScan f, RealFloat b) => f (Complex b)
phasor = fst $ lproducts (pure phaseDelta)
 where
   phaseDelta = cis (-pi / fromIntegral (sum (pure 1 :: f Int)))
```

Note that I dropped some parentheses in the `phaseDelta` definition and inlined `n`.